xdem/spatialstats.py

```diff
@@ -820,213 +1028,384 @@ def vgm_sum(h, *args):
                              p0=p0, bounds=bounds)
     # Otherwise, use a weighted fit
     else:
         # We need to filter for possible no data in the error
         valid = np.isfinite(empirical_variogram.err_exp.values)
-        cof, cov = curve_fit(vgm_sum, empirical_variogram.bins.values[valid], empirical_variogram.exp.values[valid],
+        cof, cov = curve_fit(vgm_sum, empirical_variogram.lags.values[valid], empirical_variogram.exp.values[valid],
                              method='trf', p0=p0, bounds=bounds, sigma=empirical_variogram.err_exp.values[valid])
 
-    # Provide the output function (couldn't find a way to pass this through functool.partial as arguments are unordered)
-    def vgm_sum_fit(h):
-        fn = 0
-        i = 0
-        for model in list_model:
-            if model == 'Sph':
-                fn += skg.models.spherical(h, cof[i], cof[i + 1])
-            elif model == 'Gau':
-                fn += skg.models.gaussian(h, cof[i], cof[i + 1])
-            elif model == 'Exp':
-                fn += skg.models.exponential(h, cof[i], cof[ i + 1])
-            # fn += skg.models.spherical(h, cof[i], cof[i+1])
+    # Store optimized parameters
+    list_df = []
+    i = 0
+    for model in list_models:
+        model_name = _get_scikitgstat_vgm_model_name(model)
+        # For models that expect 2 parameters
+        if model_name in ['spherical', 'gaussian', 'exponential', 'cubic']:
+            df = pd.DataFrame()
+            df = df.assign(model=[model_name], range=[cof[i]], psill=[cof[i+1]])
             i += 2
-
-        return fn
-
-    return vgm_sum_fit, cof
-
-
-def exact_neff_sphsum_circular(area: float, crange1: float, psill1: float, crange2: float, psill2: float) -> float:
+        # For models that expect 3 parameters
+        elif model_name in ['stable', 'matern']:
+            df = pd.DataFrame()
+            df = df.assign(model=[model_name], range=[cof[i]], psill=[cof[i + 1]], smooth=[cof[i+2]])
+            i += 3
+        list_df.append(df)
+    df_params = pd.concat(list_df)
+
+    # Also pass the function of sum of variogram
+    vgm_sum_fit = get_func_sum_vgm_models(df_params)
+
+    return vgm_sum_fit, df_params
+
+def _check_validity_params_vgm(params_vgm: pd.DataFrame):
+    """Check the validity of the modelled variogram parameters dataframe (mostly in the case it is passed manually)."""
+
+    # Check that expected columns exists
+    expected_columns = ['model', 'range', 'psill']
+    if not all([c in params_vgm for c in expected_columns]):
+        raise ValueError('The dataframe with variogram parameters must contain the columns "model", "range" and "psill".')
+
+    # Check that the format of variogram models are correct
+    for m in params_vgm['model'].values:
+        _get_scikitgstat_vgm_model_name(m)
+
+    # Check that the format of ranges, sills are correct
+    for r in params_vgm['range'].values:
+        if not isinstance(r, (float, np.integer)):
```

Comment:
Would that condition fail is the value was a np.float? Maybe use `np.number` rather than `np.integerer` to allow any numpy number? 
I just checked how we handle any number in geoutils here: https://github.com/GlacioHack/GeoUtils/blob/c101c3ed1e7424e0492b453b08f34aaa52b0569a/geoutils/_typing.py#L14
Same for the other conditions below.

Reply:
Solved with this: https://github.com/GlacioHack/xdem/wiki/Memo-on-Python-and-numpy-%60dtype%60